### ISSUE TITLE
Cliente ha podido hacer pedidos para el mismo día

### ISSUE THREAD

**Issue by PaulJBis** [PaulJBis/Caminofacil#332]
Un cliente ha podido hacer, el mismo día 19 de abril, un pedido con una etapa para el 19 de abril. Verificar:

-El código actual, ¿tiene solo chequeos en el lado del cliente?
-Si es así, implementar chequeos del lado del servidor de fechas.

Escribir el código de forma que, al hacer el chequeo, no sólo se puedan comprobar las fechas, sino dejar abierta también la posibilidad de chequear otras cosas. La razón es que el cliente se está quejando también de que hay otros pedidos que le están llegando sin datos de alojamiento (pero no nos ha enviado aún ninguno para que lo veamos).

**Comment by PaulJBis** [comment]
OJO: antes de subir a producción, cambiar la fecha de la variable para que no sea "3 días", tal como se explica arriba.

**Comment by PaulJBis** [comment]
Ahora mismo hay una verificación en Javascript según la cual, si es antes de las 21:00, sí podemos hacer un pedido para el día siguiente, pero si es después de las 21:00 ya no (en ese caso, la fecha más temprana para hacer un pedido sería pasado mañana).

Dicha verificación está en /app/views/public/pagina/comun/js/scripts.js. ¿Cómo interactuaría esta verificación en JS con la nueva función de validación?

Por lo que he visto, entiendo que no interactúan para nada: el código JS verifica por un lado que sean antes de las 21:00, y el código del servidor verifica que todos los datos del pedido estén bien (y si un pedido está hecho para el día siguiente, se entiende que sí está bien). Pero, ¿qué pasa si un usuario comienza a hacer un pedido hoy para mañana, lo deja varias horas y lo envía a las 23:00? ¿Lo detecta el código de servidor?

**Comment by PaulJBis** [comment]
Subido a producción el último cambio. Añadido un pedido para el 4 de mayo (4 días después de hoy). Vamos a esperar hasta después de las 21:00 y probar a ver qué pasa... ;-)

**Comment by PaulJBis** [comment]
Sobre zonas horarias: el código PHP ya está hecho para inicializar con la timezone "Europe/Madrid" al comienzo de cada ejecución (véase config.php, date_default_timezone_set()). Por tanto, las instrucciones que estamos usando deberían tenerlo en cuenta.

**Comment by PaulJBis** [comment]
Subido ahora el último merge (el último commit de la rama "validar_pedidos_servidor"), y no funciona bien. A esta hora (22:11 del 6 de mayo), si hago un pedido para el 8 de mayo, el calendario Javascript me lo admite, pero luego la validación falla. El error que da es: _"Hay etapas con pueblos de origen, pueblos de destino o fechas repetidas. Es necesario solucionarlo para finalizar el pedido."_

**Comment by PaulJBis** [comment]
Hecho otro pedido ahora, a las 0:12 del 7 de mayo. Al seleccionar una etapa para el 8 de mayo, sucede lo mismo: nos salta el error "Hay etapas con pueblos de origen, pueblos de destino o fechas repetidas. Es necesario solucionarlo para finalizar el pedido.".

Sospecho, aunque no estoy seguro, que el problema está en que usa timestamps de Unix para hacer la comparación. En otras palabras, está usando como límite:

`strtotime('+1 day');
`

Que es el timestamp de dentro de 24 horas exactas... pero luego el pedido tiene fecha simplemente del "8 de mayo de 2018", es decir, el timestamp de medianoche del "8 de mayo". ¿Es posible que sea esa la causa?

**Comment by Raimond82** [comment]
Si a las 0:12 del 7 de mayo salta error para el 8 de mayo es correcto. A partir de las 21:00 del 7 de mayo solo debería permitir hacer pedidos para el 9 de mayo. Eso es lo que yo entendí.

Te puedo confirmar que la restricción por fecha no depende de javascript, porque hay una "instrucción" que hace que javascript coja la hora del server. Mete una etiqueta en el body con la fecha del server,, que supuestamente está usando como hora js.

**Comment by PaulJBis** [comment]
Claro, a partir del 21:00 no debería permitir hacer pedidos para el día siguiente... pero las 00:12 son antes de las 21:00.

Por poner otro ejemplo: ahora mismo son las 9:12 (también “antes” de las 21:00), y el sistema no me deja hacer pedidos para mañana día 8. Esto sí que es incorrecto.

**Comment by Raimond82** [comment]
Modificado. Con strtotime utilizaba la hora actual del servidor. Ahora calcula las fechas con mktime. Si son antes de las 21:00 coge 07/05/2018 23:59:59 como limite y si son más de las 21:00 08/05/2018 23:59:59, y al validar usa las 00:00:00 del día en cuestión.

En el formulario para añadir etapas puse visible la fecha que utiliza el validador como límite para poder comprobar si cambia bien hoy a las 9.
